tchannel/container/heap.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import absolute_import
+
+
+class HeapOperation(object):
+    def cmp(self, i, j):
+        """Compare the items in position i and j of the heap.
+
+        :param i: the first item's position of the heap list
+        :param j: the second item's position of the heap list
+        :return true if items[i] < items[j], otherwise false.
+        """
+        raise NotImplementedError()
+
+    def push(self, x):
+        """Push a new item into heap"""
+        raise NotImplementedError()
+
+    def pop(self):
+        """Pop an item from the heap"""
+        raise NotImplementedError()
+
+    def swap(self, i, j):
+        """swap items between position i and j of the heap"""
+        raise NotImplementedError()
+
+
+def init(h):
+    # heapify
+    n = len(h)
+    for i in range(n/2 - 1, -1):
+        down(h, i, n)
+
+
+def push(h, x):
+    h.push(x)
+    up(h, len(h)-1)
+
+
+def pop(h):
+    n = len(h) - 1
+    h.swap(0, n)
+    down(h, 0, n)
+    return h.pop()
+
+
+def remove(h, i):
+    n = len(h) - 1
+    if n != i:
+        h.swap(i, n)
+        down(h, i, n)
+        up(h, i)
+
+    return h.pop()
+
+
+def fix(h, i):
+    down(h, i, len(h))
+    up(h, i)
```

Comment:
It will walk up the tree without making any changes. It' ll check that `parent(i)` is less than the new `i`, and that `parent(parent(i))` is less than that, and so on until it reaches the root node. All these conditions will be true because of previously explained reasons. These comparisons are very easy to avoid by checking if `down()` sifted the value down, and if so, don't call `up()`.